### DISCUSSION TITLE
3 new mods: "Renaissance", "Slice" and "Repeat" ─ enhance map practicing, the osu!lazer way!

### DISCUSSION THREAD

**Discussion by Speykious** [ppy/osu#14629] [Ideas] [answered]
# The current observed issues that lead me to the idea

For a long time, there have been no in-game option in osu!lazer (nor in osu!stable AFAIK) meant to practice a map, apart from playing it over and over again.
Meaning, if you want to practice a specific map (for various reasons: you like it, you wanna farm, you wanna get an SS somewhere, someone gave you a challenge, etc.), you have very few choices:
- play it over and over again,
- using the Sudden Death or Perfect mod at one point,
- make a sliced copy of the map itself, which I don't know if it has actually ever been done due to how tedious it is to do so.

Some other intense games where you are bound to time and have to finish a level have some way to practice them: Geometry Dash for example has a Practice mode where you get checkpoints once in a while so that you respawn mid-way through when you die.
So I think that it is a must for osu! to have some ergonomic way of practicing.

# Enhance map practicing: 3 new mods

## Renaissance

This mod acts very similarly to Geometry Dash's Practice Mode (let's call it GDPM).
- It will regularly give you checkpoints, and put you back to the latest one when you die.
- Ideally, the checkpoints will be aligned with the beat of the map.
- The small pause menu when pressing Escape will have options to go back to previous checkpoints, when you're in a situation where you can't get your rhythm back.
- Also ideally, the song of the map plays in the background, unlike GDPM.
- The regularity of the checkpoints is customizable.
- Maybe add a keybind for a manual checkpoint feature, like the spacebar or something.
- Since it essentially allows you to cheat death, it should divide the score the same way No Fail, Easy and Half Time mods do. I would suggest 0.50x as a factor.

## Slice

This mod lets you play a specific slice of the map ─ from a timestamp A to a timestamp B.
- You can customize when the slice starts and when it ends, ideally measured in beats rather than seconds.
- The music in the background fades in (unless at the actual start of the map), maybe as a toggleable option.
- Ideally, in terms of star rating, it would have the same effect as selecting the specific hit objects of the map in the editor and make a new map out of it.

## Repeat

This mod lets you play the same map multiple times.
- You can customize how many times you play the map, from 2 to some threshold.
- You can also customize the length of the breaks between each repeat.

# All intercompatible

My 3 mod propositions up there are meant to all be compatible with each other. You could for example decide to practice a specific slice of the map 5 times with checkpoints. Even adding the Perfect mod into the mix to restart at the latest checkpoint at the first slightest error.
They should also be compatible with every other mod: taking the same example again, you could decide to practice it with Hidden, Hard Rock, Traceable or any other currently existing combination.

# Edit

Contrary to what the first answer suggests, **this is not a duplicate**. The goal of this discussion post is not to label my ideas as entirely new, because it is obvious that after 14 years of lack of practice mod, someone was gonna suggest it out before me. Instead, it is to make the already existing idea **more fleshed out**, with more details and regards to how osu!lazer works.

**Comment by bdach** [comment, marked as answer]
Pretty much none of the ideas given are new.

https://github.com/ppy/osu/discussions/13079
https://github.com/ppy/osu/discussions/13076
https://github.com/ppy/osu/discussions/13013

**Reply by Speykious** [reply]
Now of course, I expected a practice mod to have already been mentioned.
However:
- discussion #13079 introduces the idea of a "playback mod" as follows:
  > "a mod that will randomly jump back to part of the timeline when playing a song"
  
  I give more details, such as how it's supposed to be customized.
- discussion #13076 introduces the idea of a completely different mod that has nothing to do with my "Renaissance" mod idea. It instead essentially talks about a mod that adds a cursor guide to your gameplay so that you can compare yours to an autoplay in real time.
- discussion #13013, similarly to discussion #13079, doesn't give much detail, though there are still more compared to it:
  > Add feature maybe like training mod or something like that, specify the start and end point/time on the map that player wants to try so players can play that specific part without creating new difficulty

I still give more information, and also talk about their composability, which is why I made one discussion for the 3 mods, because there is something to be discussed about how they interact with each other.

**Reply by LittleEndu** [reply]
You have to realize you can't speak like you are in a position of authority. You can't dismiss concerns raised by one of the collaborators of osu!lazer. You can't talk about a practice mod like it's your idea no matter how you might split and reword it. 

The fact is when practice mod is implemented people already expect it to do everything you described and iff it doesn't should you start requesting improvements. This entire idea is now fragmented over like 7 different posts and none of them are going to be looked at by the person who is actually going to implement the mod.

**Reply by Speykious** [reply]
I apologize for the miscommunication. I never intended to "speak like I am in a position of authority", nor ─ which is the reason for the disclaimer that I added minutes ago ─ that it's "my idea".
I also apologize that I contributed to the further fragmentation of the posts talking about the general idea of a practice mod. If there's a way to delete this discussion, I'd gladly ask for it.

*Still* however, I have a question: why would every post be disregarded by the implementor?

**Reply by LittleEndu** [reply]
Not disregarded as in when the practice mod is implemented it will probably satisfy all those proposal posts. It just makes more sense to have something implemented first before going into all the details of how the feature should work. Sorry I can't explain it better, but osu! has few developers and they need to manage their time the best they can to actually get anywhere.

**Reply by Speykious** [reply]
I see.

**Reply by gagahpangeran** [reply]
As a side note : you are welcome to implement this mod and start with something simple.

**Reply by peppy** [reply]
Ideas are cheap. Implementation specifics will be governed by myself for something like this (I'm a bit fussy with UX when it comes to new features). But I agree that a start would be to get something simple implemented and we can work from there.

**Reply by Speykious** [reply]
Thanks for your comments. If I get any free time, I might look more thoroughly into the source code.
